tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -733,351 +714,312 @@ getParamParentNameOffset(const ValueDecl *VD, SourceLoc Cursor) {
-                                  SourceLoc CursorLoc,
-                  ArrayRef<RefactoringInfo> KownRefactoringInfoFromRange,
-                                  SwiftLangSupport &Lang,
-                                  const CompilerInvocation &Invoc,
-                                  std::string &Diagnostic,
-                      ArrayRef<ImmutableTextSnapshotRef> PreviousASTSnaps,
-                  std::function<void(const RequestResult<CursorInfoData> &)> Receiver) {
-  if (AvailableAttr::isUnavailable(VD)) {
-    Diagnostic = "Unavailable in the current compilation context.";
-    return false;
-  }
-
-  SmallString<64> SS;
-  auto BaseType = findBaseTypeForReplacingArchetype(VD, ContainerTy);
+struct DeclInfo {
+  const ValueDecl *VD;
+  Type ContainerType;
+  bool IsRef;
+  bool IsDynamic;
+  ArrayRef<NominalTypeDecl *> ReceiverTypes;
+
+  /// If VD is a synthesized property wrapper backing storage (_foo) or
+  /// projected value ($foo) of a property (foo), the property instead.
+  /// Otherwise, VD.
+  const ValueDecl *Underlying = nullptr;
+  bool Unavailable = true;
+  Type BaseType;
   bool InSynthesizedExtension = false;
-  if (BaseType) {
-    if (auto Target = BaseType->getAnyNominal()) {
-      SynthesizedExtensionAnalyzer Analyzer(
-          Target, PrintOptions::printModuleInterface(
-                      Invoc.getFrontendOptions().PrintFullConvention));
-      InSynthesizedExtension = Analyzer.isInSynthesizedExtension(VD);
-    }
-  }
 
-  unsigned NameBegin = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    SwiftLangSupport::printDisplayName(VD, OS);
-  }
-  unsigned NameEnd = SS.size();
+  DeclInfo(const ValueDecl *VD, Type ContainerType, bool IsRef, bool IsDynamic,
+           ArrayRef<NominalTypeDecl *> ReceiverTypes,
+           const CompilerInvocation &Invoc)
+      : VD(VD), ContainerType(ContainerType), IsRef(IsRef),
+        IsDynamic(IsDynamic), ReceiverTypes(ReceiverTypes) {
+    if (VD == nullptr)
+      return;
 
-  // If VD is the syntehsized property wrapper backing storage (_foo) or
-  // projected value ($foo) of a property (foo), use that property's USR instead
-  // so that a rename refactoring renames all three (foo, $foo, and _foo).
-  const ValueDecl* OriginalProperty = VD;
-  if (auto *VarD = dyn_cast<VarDecl>(VD)) {
-    if (auto *Wrapped = VarD->getOriginalWrappedProperty())
-        OriginalProperty = Wrapped;
-  }
+    Underlying = VD;
+    if (auto *VarD = dyn_cast<VarDecl>(VD)) {
+      if (auto *Wrapped = VarD->getOriginalWrappedProperty())
+        Underlying = Wrapped;
+    }
 
-  unsigned USRBegin = SS.size();
-  {
-    llvm::raw_svector_ostream OS(SS);
-    SwiftLangSupport::printUSR(OriginalProperty, OS);
-    if (InSynthesizedExtension) {
-        OS << LangSupport::SynthesizedUSRSeparator;
-        SwiftLangSupport::printUSR(BaseType->getAnyNominal(), OS);
+    Unavailable = AvailableAttr::isUnavailable(Underlying);
```

Comment:
This used to early-exit on unavailable decls, and it used to check VD not Underlying.  Are these intentional changes?  If so, what's the motivation?  If not, should we change it back?

Reply:
They were both intentional, but:
  - Early exit: I had a quick look at the implementations of getting the base type/checking for synthesized extension and they seemed okay even when unavailable. In the interests of fewer changes I'll add an early exit though.
  - Underlying vs VD: My thinking here was that we want the same behaviour for foo, $foo, and _foo. As it turns out though, $foo and _foo aren't marked as unavailable. I'll change this back and also open a ticket to allow unavailable when we have a location, since that's where this matters.